Import activity data into lessons

bethanyaconnor opened the pull request:
This module was what I started work on and it turned out to work fairly well. It's a big wall of code but it is self-contained.

This script successfully runs on CSF, CSD, and CSP 2020. There are definitely still kinks, but I figure this is something we can work off of.

What this script attempts to do:
- Separate tips and their text into their own activity sections
  - There are some tips that don't have a place for the icon in CB. These activity sections don't have a description and are put at the end
  - There are some tips that have two places referring to them. I chose the first one and the second one is rendered without the tip icon.
- Find remarks (`say` blocks) and put them in their own activity section
- A first attempt as slides
  - In CB, this syntax just renders the icon. There are places where this will just not be supported ([example](https://curriculum.code.org/csp-20/unit1/1/#tip_tip-0))

What is a future improvement I'd like to make
- All levels are just stuck in their own activity at the end. I had some issues with them getting reordered and decided to put that work down for a bit.
- Activity sections never have titles. We could try a best effort to look for a heading at the beginning of the markdown or something in the future.
- General improvements on the regexes and heuristics used here.

What I'm unsure on
- How do progressions work with the new lesson plans? They disappeared when I run this script, but I'm not sure how to properly represent them
- Are we supporting the `[guide][/guide]` syntax from CB? Currently, it's just being ignored.

One sample [lesson](https://curriculum.code.org/csd-20/unit3/1/), side by side:

![Screenshot 2020-11-11 at 12 50 32 PM](https://user-images.githubusercontent.com/46464143/98863234-ef746900-241c-11eb-8333-360c218a7e3e.png)
![Screenshot 2020-11-11 at 12 50 55 PM](https://user-images.githubusercontent.com/46464143/98863244-f307f000-241c-11eb-8031-db94f18cbd47.png)
![Screenshot 2020-11-11 at 12 51 24 PM](https://user-images.githubusercontent.com/46464143/98863253-f56a4a00-241c-11eb-8fed-5e9575bd940f.png)

## Testing story

a lot of reseed and rerun locally

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

dmcavoy commented:
So excited to see this coming together Bethany. I haven't looked at the code yet but responding to your overview below.

>   * There are some tips that don't have a place for the icon in CB. These activity sections don't have a description and are put at the end

I think this is just user error and I would think that we should just guess the icon location would be where the tip is.

>   * There are some tips that have two places referring to them. I chose the first one and the second one is rendered without the tip icon.

Can you share an example of this. I'm not clear what you are saying here.

>   * In CB, this syntax just renders the icon. There are places where this will just not be supported ([example](https://curriculum.code.org/csp-20/unit1/1/#tip_tip-0))

We should discuss this with curriculum and see if they are ok with this not being supported.

> * Activity sections never have titles. We could try a best effort to look for a heading at the beginning of the markdown or something in the future.

In your example I see a bunch of headers I would have thought to be Activity Section Headers such as 'The Entertainment Problem'. Can we look for anything that  is '### The Entertainment Problem' and make that the title of an activity section with everything following it until a new activity section is needed?

> What I'm unsure on
> 
> * How do progressions work with the new lesson plans? They disappeared when I run this script, but I'm not sure how to properly represent them

So levels that were in a progression together before should now be in an activity section together where the activity section title is the progression name.

Example old DSL:
level 'CSD U2 link demo_2020', progression: 'Exploration'
level 'CSD U2 making links_2020', progression: 'Skill Building'
level 'CSD U2 naming links_2020', progression: 'Skill Building'
level 'CSD U2 naming pages_2020', progression: 'Skill Building'
level 'CSD U2 nav bar_2020', progression: 'Practice'

New Activity Sections:

Title: Exploration
Levels: CSD U2 link demo_2020

Title: Skill Building
Levels: CSD U2 making links_2020, CSD U2 naming links_2020, CSD U2 naming pages_2020

Title: Practice
Levels: CSD U2 nav bar_2020


> * Are we supporting the `[guide][/guide]` syntax from CB? Currently, it's just being ignored.

Last conversation with the curriculum team about this they don't think that teachers know what it means and so they didn't seem to be suggesting that it needed to mover over. So I would ignore for now and we can bring it up again.

dmcavoy commented on dashboard/app/helpers/lesson_import_helper.rb:
```diff
@@ -0,0 +1,213 @@
+# Helper module for importing data from curriculumbuilder
+module LessonImportHelper
+  # Lockable lessons don't need to be merged with curriculumbuilder, but we do
```
There are CSP lessons where a lockable lesson has a lesson plan. https://www.codecurricula.com/csp-20/unit4/15/

bethanyaconnor replied:
This should now be addressed between Dave's recent changes and 75f2de5e0897db75e0c4597f5c889d55bef85289.

dmcavoy commented on dashboard/app/helpers/lesson_import_helper.rb:
```diff
@@ -0,0 +1,213 @@
+    activity_section.seeding_key = SecureRandom.uuid
+    activity_section.position = 1
+    activity_section.lesson_activity = lesson_activity
+    activity_section.save!
```
do you need to call lesson_activity.save! too somewhere in here?

bethanyaconnor replied:
I think `lesson_activity.save` is automatically called when it's set as the lesson_activity of the activity_section. Not 100% sure though...

dmcavoy commented on dashboard/app/helpers/lesson_import_helper.rb:
```diff
@@ -0,0 +1,213 @@
+
+  # https://github.com/code-dot-org/remark-plugins/blob/master/src/tiplink.js
+  def self.find_tip_links(markdown)
+    regex = /^([\w-]+)!!! ?([\w-]+)(?:<!-- place where you'd like the icon -->)?(.+?)$/
```
have you run into any where the comment was removed? I used to remove the comment sometimes

bethanyaconnor replied:
If you're thinking of the `<!-- place where you'd like the icon -->` bit, yeah it is removed sometimes. The `?` at the end makes it optional, so this should match whether or not it's there. I have to write it out explicitly like this so that we can get the capture after it is correct.

dmcavoy replied:
Ahhh . I'm really bad with regular expressions. That makes sense

bethanyaconnor replied:
I should comment these better - these were created after looking at a number of sources plus using a regex checker, so they could use a bit of comments. I'll make a pass.

dmcavoy commented on dashboard/app/helpers/lesson_import_helper.rb:
```diff
@@ -0,0 +1,213 @@
+  end
+
+  def self.find_slides(markdown)
+    regex = /^slide!!!slide-\d+(?:<!-- place where you'd like the icon -->)?/
```
Also wondering here if the comment is not there sometimes

bethanyaconnor commented:
Super helpful thanks Dani!

> > There are some tips that don't have a place for the icon in CB. These activity sections don't have a description and are put at the end

> I think this is just user error and I would think that we should just guess the icon location would be where the tip is.

That makes sense. I'd like to add that to the things to improve on in the future if that's alright though. Putting these leftover tips at the end was easier so trying to figure out their location will require a bit of refactoring.

> > There are some tips that have two places referring to them. I chose the first one and the second one is rendered without the tip icon.

> Can you share an example of this. I'm not clear what you are saying here.

https://curriculum.code.org/csd-20/unit2/3 has an example. There are two places that refer up to the same tip about pair programming. The way I handled this was to have the first place actually have the tip attached and the second one to just be there with no tip icon.

> In your example I see a bunch of headers I would have thought to be Activity Section Headers such as 'The Entertainment Problem'. Can we look for anything that is '### The Entertainment Problem' and make that the title of an activity section with everything following it until a new activity section is needed?

I was just playing around with editing these lessons to track down a bug and that makes a lot of sense. I'll play with that a bit here but if it ends up being too challenging, I'll defer to a future PR.

> So levels that were in a progression together before should now be in an activity section together where the activity section title is the progression name.

That makes a lot of sense! I'll fix that for this PR. I'd still like to play around with where the levels are put in the lesson plan in a future PR but I think it makes sense to at least have the progressions right on the first pass.

davidsbailey commented on bin/oneoff/import_unit_details.rb:
```diff
@@ -100,7 +100,10 @@ def main(options)
       lesson.update_from_curriculum_builder(cb_lesson)
+      log("update lesson #{lesson.id} with cb lesson data: #{cb_lesson_data.to_json[0, 50]}...")
     end
+    script.lessons.select(&:lockable).each(&:update_from_curriculum_builder)
```
Ahh, it looks like this is needed because otherwise update_from_curriculum_builder doesn't get called for lockable lessons without lesson plans, is that correct? it kind of stands out because (1) some lockable lessons (such as in CSP) were already included in lesson_pairs, and (2) no arguments are passed to update_from_curriculum_builder. The combination of these two seems like it might be bad, because a lockable lesson might first receive some cb data passed to update_from_curriculum_builder, but then it will receive empty data on the second call here.

One idea would be to filter out lockable lessons which exist in lesson_pairs before making the call to update_from_curriculum_builder on this line.

bethanyaconnor replied:
Yep. This came up because fixing the level positions would fail on non-CSF courses. Though I didn't realize that some lockable lessons have lesson plans, so I'll need to update the logic here.

davidsbailey commented on dashboard/app/helpers/lesson_import_helper.rb:
```diff
@@ -0,0 +1,213 @@
+    @lesson_activity.save!
+    @lesson_activity.reload
+    @lesson_activity.activity_sections = [create_activity_section_with_level_ranges(0, levels.count - 1, levels)]
+    activities.push(@lesson_activity)
```
I'm a bit concerned about `@lesson_activity` as a class variable. it looks you should be able to:
1. declare it as `lesson_activity = nil` at the beginning of the method
2. pass it as a parameter to create_activity_section_with_level_ranges

bethanyaconnor replied:
Yeah, I got bitten by this bad practice already haha. I'll move away from it.

davidsbailey commented on dashboard/app/helpers/lesson_import_helper.rb:
```diff
@@ -0,0 +1,213 @@
+# Helper module for importing data from curriculumbuilder
+module LessonImportHelper
```
I don't want to create a bunch of unnecessary work here, but I would caution against using rails helpers since they make it easy to violate encapsulation. If it is not too much work, I would recommend moving this file to `dashboard/lib/services/` and then requiring it directly . As another benefit, in the future we could move update_from_curriculum_builder into that file, excluding all of this code from our rails application entirely and only loading it into our import script.

bethanyaconnor replied:
That sounds good to me. I'll move it over.

davidsbailey reviewed (commented):
Awesome progress, Bethany! I had to think to figure out which image was before and which was after 😁  Some initial comments inline.

dmcavoy commented:
Bethany - Just following up to say that waiting for a later PR for any of the suggestions I gave is totally fine.

dmcavoy commented:
@bethanyaconnor In https://www.codecurricula.com/csd-20/unit2/3/. I think that there is actually two tips with two tip icons in the Main Activity. They are just numbered the same so thats confusing. We might have to do something to check if there are more than one tip/tip icon with the same numbering.

bethanyaconnor commented:
@dmcavoy ooo I see -- you're right. I'll see if I can do something a bit cleverer to cover this case.

bethanyaconnor commented:
To recap what I've fixed here:
- Levels are now put into the correct progressions. They're still in their own activity though.
- Moved this utility to `dashboard/lib/services/`
- Removed all the class variables
- Use CB data for lockable lessons and just move the levels for lockable lessons without a CB counterpart
- Some more general other improvements

I'd like to get this PR in a place to merge it and follow up with smaller PRs to:
- Import any outstanding Lesson fields
- More work on turning headings into activity section titles
- Try to put tips that don't have a corresponding paragraph in the correct location, instead of at the end
- More general improvements/bug fixes

Also import objectives, which will likely be related to this work.

davidsbailey commented on dashboard/lib/services/lesson_import_helper.rb:
```diff
@@ -0,0 +1,260 @@
+      elsif match[:type] == 'remark'
+        activity_section = create_activity_section_with_remark(match[:match])
+      else
+        #activity_section = ActivitySection.new(description: match[:substring].strip)
```
remove dead code?

davidsbailey commented on dashboard/lib/services/lesson_import_helper.rb:
```diff
@@ -0,0 +1,260 @@
+
+  def self.find_tips(markdown)
+    # See https://github.com/code-dot-org/remark-plugins/blob/master/src/tip.js
+    regex = /^!!! *?([\w-]+)(?: "(.*?)")?(?: <(.*?)>)?(?:[\s]+$)+([\d\D]+?)(?=(^\S|^$))/
```
I love the new comments on regexes, but this one's still a little hard for me to understand even after looking at the other file. I think it could help to include some very general context here, something like 
```
# Finds teaching tips or content tips, such as those starting with `!!! say` or `!!! discuss`
```
(my actual text here is unlikely to be accurate)

davidsbailey commented on dashboard/lib/services/lesson_import_helper.rb:
```diff
@@ -0,0 +1,260 @@
+  def self.create_activity_sections(activity_markdown)
+    tip_matches = find_tips(activity_markdown).select {|m| m[1] != 'say'}.map {|m| {index: activity_markdown.index(m[0]), type: 'tip', match: m, substring: m[0]}}
+    tip_link_matches = find_tip_links(activity_markdown).map {|m| {index: activity_markdown.index(m[0]), type: 'tiplink', match: m, substring: m[0]}}
+    remark_matches = find_remarks(activity_markdown).map {|m| {index: activity_markdown.index(m[0]), type: 'remark', match: m, substring: m[0]}}
```
sorry for the half-baked investigation here, but it is a bit hard to see whether there is overlap between `find_remarks` and `find_tips`, such as whether they might both match `!!! say`, for example. Perhaps improving the comments on `find_tips` will help clarify whether this is true or not.

bethanyaconnor replied:
Yeah `find_tips` will also find remarks. I could definitely do some cleanup to fix this up, but I might save that for a followup PR. (I'll improve comments in this PR though).

davidsbailey reviewed (approved):
Nice updates, Bethany! I haven't gone super deep here on the details on this one, since I think Dani is the expert here. Feel free to respond to feedback in a later PR if that's easier.

dmcavoy commented on dashboard/lib/services/lesson_import_helper.rb:
```diff
@@ -0,0 +1,279 @@
+
+    # Create a map of tip key -> an array of tip matches
+    # Sometimes, a key is used multiple times in the same activity. To handle
+    # this we'll, just pair them with the tip links in order.
```
nice!

dmcavoy commented on dashboard/lib/services/lesson_import_helper.rb:
```diff
@@ -0,0 +1,279 @@
+    end
+
+    # If there are any tips that didn't have a match, put them at the end
+    # TODO ideally, we'd put them in the spot they were written.
```
It would be interesting for this to print out what lessons this is happening on so we can investigate it

bethanyaconnor replied:
I'm adding a puts statement -- it looks like this happens in a fair number of lessons. Some might be bugs in the script to be fair but not all (see [this lesson](https://curriculum.code.org/csf-20/coursea/2/))

dmcavoy commented on dashboard/lib/services/lesson_import_helper.rb:
```diff
@@ -0,0 +1,279 @@
+    activity_section
+  end
+
+  # No levels, no tips, just markdown
```
so this means right now if we find a header we are making it its own activity section and not combining it in an activity section with some of the stuff below it?

bethanyaconnor replied:
Yeah currently :/ Now that I'm thinking through this more, this might actually be a relatively straightforward fix though, so I'll try to fix this today in a separate PR.

davidsbailey commented:
I noticed the following issue after pulling this branch down locally. starting from a clean slate, importing csp1-2020 succeeds once but then fails if I run it again:

1. `[development] dashboard > Script.find_by_name('csp1-2020').destroy`

2. `Dave-MBP:~/src/cdo/dashboard (lesson-import-helper)$ rake seed:single_script SCRIPT_NAME=csp1-2020`

3. `Dave-MBP:~/src/cdo (lesson-import-helper)$ bundle exec ./bin/oneoff/import_unit_details.rb -l -u csp1-2020 -v`

succeeds

4. `Dave-MBP:~/src/cdo (lesson-import-helper)$ bundle exec ./bin/oneoff/import_unit_details.rb -l -u csp1-2020 -v`

<img width="1188" alt="Screen Shot 2020-11-17 at 7 45 33 PM" src="https://user-images.githubusercontent.com/8001765/99480727-85246280-290d-11eb-9cb0-fab97631975e.png">

I'm not sure how important this is, but assuming there is more work to do on import, it might save us some time if we can make the import script idempotent. I may try to take a crack at this, but if you have a pro tip on why this might be happening please let me know :-) It appears to only happen in CSP, probably because those are the only places we have lesson plans for lockable lessons.

bethanyaconnor commented:
@davidsbailey interesting! I've definitely rerun the script a few times and I don't think it's idempotent because of the script levels but I don't think I've ever seen it crash 🤔

bethanyaconnor commented:
Nevermind I'm seeing the same crash now. Looking...

bethanyaconnor commented:
I removed setting the `relative_position` on the lesson on import in 1a28a2d because I _think_ that the studio lesson should already have that. This fixes the crash on rerun, though again, I don't think the script level changes are idempotent.

davidsbailey commented:
Thanks @bethanyaconnor ! "idempotent" was a too strong a choice of words on my part. "safe to rerun without crashing" is what I was after. 

For now, I'll use your trick of disabling the create_lesson_activities call (once it has been run initially for each script) if I need to make any tweaks to importing resources or anything else besides lesson activities.

davidsbailey commented:
> I removed setting the `relative_position` on the lesson on import in [1a28a2d](https://github.com/code-dot-org/code-dot-org/commit/1a28a2d27c1a29a9c83aa8d242527dec82ce5c6d) because I _think_ that the studio lesson should already have that. This fixes the crash on rerun, though again, I don't think the script level changes are idempotent.

Oooh, yeah I think that explains it, and also I think it might have broken other things if we let that change go through. csp1-2020 for example has two lockable lessons and only the second one has a corresponding lesson plan in CB, so we definitely wouldn't want to changes its relative position. If i run your import script before the commit you mentioned, for example, then this link breaks: http://localhost-studio.code.org:3000/s/csp1-2020/lockable/2/puzzle/1/page/1 Thanks for following up on this!
